src/indexes/prefix.rs

```diff
@@ -0,0 +1,636 @@
+#![cfg(feature = "iterator")]
+use core::fmt;
+use cosmwasm_std::storage_keys::to_length_prefixed_nested;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+
+use cosmwasm_std::{Order, Record, StdResult, Storage};
+use std::ops::Deref;
+
+use crate::bound::RawBound;
+use crate::de::KeyDeserialize;
+use crate::iter_helpers::{concat, deserialize_kv, deserialize_v, trim};
+use crate::keys::Key;
+use crate::{Bound, PrimaryKey};
+
+type DeserializeVFn<T> = fn(&dyn Storage, &[u8], Record) -> StdResult<Record<T>>;
+
+type DeserializeKvFn<K, T> =
+    fn(&dyn Storage, &[u8], Record) -> StdResult<(<K as KeyDeserialize>::Output, T)>;
```

Comment:
More of a comprehension question on my side: Originally I was wondering whether it would be possible to change these to something like
```rust
type DeserializeVFn<T> = fn(&dyn Storage, &[u8], Vec<u8>) -> StdResult<T>;
type DeserializeKFn<T> = fn(&dyn Storage, &[u8], Vec<u8>) -> StdResult<<K as KeyDeserialize>::Output>;
```
but if I understand correctly, that would not work because the index needs to be able to produce a key from the value, right?

Reply:
Yes. When I worked on this I attempted that, but I couldn't get it working for exactly that reason.